service-mirror.go

```diff
@@ -34,110 +42,56 @@ func main() {
+	svcWatcher := NewServiceWatcher(*ctx, log, globalSvcNs)
 
-	if err != nil {
-		log.Errorf("Unable to get running services for filter : %s, Error: %s", svcFilter.String(), err.Error())
-	}
-
-	//Map of Mirrored services with targetCluster as key and responding service
-	mirroredService := make(map[string][]corev1.Service)
-	log.Debugln("All running Services :")
-	for _, svc := range mirroredServices.Items {
-		//Only gather service which doesnt have label mirror.linkerd.io/headless-mirror-svc-name, as this are the parent services
-		targetClusterame := svc.GetLabels()["mirror.linkerd.io/cluster-name"]
-		if _, ok := mirroredService[targetClusterame]; !ok {
-			mirroredService[targetClusterame] = []corev1.Service{}
-		}
-		mirroredService[targetClusterame] = append(mirroredService[targetClusterame], svc)
+	watcher := &Watcher{}
+	watcher.informers = append(watcher.informers, svcWatcher.informer)
 
+	//Run all the informers
+	for _, informer := range watcher.informers {
+		go informer.Run(stopCh)
 	}
 
-	//Iterate over map of targetCluster -> headlessService and check if global service exists.
-	for tgCluster, mirrorService := range mirroredService {
-		log.Debug("Key=", tgCluster)
-		for _, hdlSvc := range mirrorService {
-			hdlSvc = *hdlSvc.DeepCopy()
-			//Check if global/aggregator service exists
-			log.Debugf("Svc from cluster = %s", hdlSvc.Name)
-			globalSvcName := strings.Split(hdlSvc.Name, fmt.Sprintf("-%s", tgCluster))[0]
-			globalSvcName = globalSvcName + "-global"
-			log.Infof("Global svc Name = %s ", globalSvcName)
-			_, err := client.CoreV1().Services("").Get(context.Background(), globalSvcName, metav1.GetOptions{})
-
-			//If global service doesnt exist cerate it
-			if !apiError.IsAlreadyExists(err) {
-
-				/*
-					- Spin up the new global service with cardinal index as x-global,
-					Which will be aggregator for mirrored services from target cluster x-target0, x-target1
-				*/
-
-				svcMeta := &metav1.ObjectMeta{
-					Name:      globalSvcName,
-					Namespace: hdlSvc.Namespace,
-					Labels: map[string]string{
-						"mirror.linkerd.io/global-mirror-for": tgCluster,
-					},
-				}
-
-				svcSpec := &corev1.ServiceSpec{
-					ClusterIP: "None",
-					Ports:     hdlSvc.Spec.Ports,
-				}
-				globalService := &corev1.Service{
-					ObjectMeta: *svcMeta,
-					Spec:       *svcSpec,
-				}
-				//Create clientSet to create Service
-				defaultCreateOptions := metav1.CreateOptions{}
-				_, err := client.CoreV1().Services(hdlSvc.Namespace).Create(context.Background(), globalService, defaultCreateOptions)
-				if !apiError.IsAlreadyExists(err) {
-					log.Infof("Service already exists with, name=%s", globalSvcName)
-				}
-
-			}
-			log.Infof("Global Service with Name: %s  should be existing", globalSvcName)
+	//Make sure cache is synced.
+	for _, informer := range watcher.informers {
+		if !cache.WaitForCacheSync(stopCh, informer.HasSynced) {
```

Comment:
Should we sync before we run?

Reply:
I derived this from one of example : https://www.cncf.io/blog/2019/10/15/extend-kubernetes-via-a-shared-informer/

From my understanding, the way this works is because of shared channel. In waitforcachesync we use the same channel, which we pass while running informer. So  waitforcachesync can block the go routine. Till the informer hasn't synced the cache.